### ISSUE TITLE
Transfer learning

### ISSUE THREAD

**Issue by anovoam** [HuguesTHOMAS/KPConv#43]
Hi @HuguesTHOMAS 

Thank you very much for your awesome work. 

I saw the instructions for creating a new dataset for training your code on it. However, I would like to use it to perform transfer learning instead. I would like then to use the learned weights of the network but train the input and output layers.

I have trained and tested the Semantic3D model. However, I am not able to actually find where the "model" is in order to freeze the intermediate layers and do the other modifications for using my data. I was wondering if you could give me some advice about this.

Many thanks!

**Comment by HuguesTHOMAS** [comment]
Hi @anovoam,

The model weights are loaded when you call the initialization method of the `ModelTrainer` class. This method is defined here : 

https://github.com/HuguesTHOMAS/KPConv/blob/935a2bd75775b07eda8cecbacf2d91b2dd7cbf8d/utils/trainer.py#L55-L103

First, In this code you can see a parameter `restore_snap` gives a previous model to load. From line 96 to 103, you can choose the variables that are loaded from the previous model. The other variables are simply initialized as if it was a new model. 

Then, if you want to freeze some of the intermediate layers, its is better to act on the optimizer than on the variables. The optimizer is defined in the methods `add_train_ops`, which is called during the init methods.

https://github.com/HuguesTHOMAS/KPConv/blob/935a2bd75775b07eda8cecbacf2d91b2dd7cbf8d/utils/trainer.py#L114-L123

You can modify the line 123 with a list of the variable the optimizer is going to update. For example :

    # Training step op
    new_trainable_vars = []
    for var in tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES, scope='KernelPointNetwork'):
        layer_name = var.name.split('/')[1]
        if layer_name not in your_list_of_frozen_layers:
            new_trainable_vars += [var]
    gvs = optimizer.compute_gradients(model.loss, var_list=new_trainable_vars)

I hope this is clear enough.
Best,
Hugues

**Comment by anovoam** [comment]
This worked great. Thank you!

**Comment by Connect2Aditya** [comment]
Hi @anovoam 
Can you share your final trainer.py file that you used for this task.
Thanks.

**Comment by Connect2Aditya** [comment]
Hi @HuguesTHOMAS 

Can you name some variables the optimizer is going to update in line 123.
